libraries/state_history/include/eosio/state_history/log.hpp

```diff
@@ -68,818 +66,405 @@ static const uint16_t ship_feature_pruned_log = 1;
-         pos += (sizeof(state_history_log_header) + payload_size + sizeof(uint64_t));
-      }
-   }
+   fc::random_access_file::device device;
+   uint64_t                       compressed_data_offset;
+   uint64_t                       compressed_data_size;
+   std::optional<uint64_t>        uncompressed_size;
 };
 
-// directly adapt from boost/iostreams/filter/counter.hpp and change the type of chars_ to uint64_t.
-class counter  {
+class state_history_log {
 public:
-    typedef char char_type;
-    struct category
-        : bio::dual_use,
-          bio::filter_tag,
-          bio::multichar_tag,
-          bio::optimally_buffered_tag
-        { };
-    explicit counter(uint64_t first_char = 0)
-        : chars_(first_char)
-        { }
-    uint64_t characters() const { return chars_; }
-    std::streamsize optimal_buffer_size() const { return 0; }
-
-    template<typename Source>
-    std::streamsize read(Source& src, char_type* s, std::streamsize n)
-    {
-        std::streamsize result = bio::read(src, s, n);
-        if (result == -1)
-            return -1;
-        chars_ += result;
-        return result;
-    }
-
-    template<typename Sink>
-    std::streamsize write(Sink& snk, const char_type* s, std::streamsize n)
-    {
-        std::streamsize result = bio::write(snk, s, n);
-        chars_ += result;
-        return result;
-    }
-private:
-    uint64_t chars_;
-};
-
-} // namespace detail
+   using non_local_get_block_id_func = std::function<std::optional<chain::block_id_type>(chain::block_num_type)>;
 
-class state_history_log {
- private:
-   const char* const       name = "";
-   state_history_log_config _config;
+   static std::optional<chain::block_id_type> no_non_local_get_block_id_func(chain::block_num_type) {
+      return std::nullopt;
+   }
+private:
+   std::optional<state_history::prune_config> prune_config;
+   const non_local_get_block_id_func          non_local_get_block_id;
 
-   // provide exclusive access to all data of this object since accessed from the main thread and the ship thread
-   mutable std::mutex      _mx;
-   fc::cfile               log;
-   fc::cfile               index;
-   uint32_t                _begin_block = 0;        //always tracks the first block available even after pruning
-   uint32_t                _index_begin_block = 0;  //the first block of the file; even after pruning. it's what index 0 in the index file points to
-   uint32_t                _end_block   = 0;
-   chain::block_id_type    last_block_id;
+   fc::random_access_file       log;
+   fc::random_access_file       index;
+   uint32_t                     _begin_block       = 0;  //always tracks the first block available even after pruning
+   uint32_t                     _index_begin_block = 0;  //the first block of the file; even after pruning. it's what index 0 in the index file points to
+   uint32_t                     _end_block         = 0;
```

Comment:
yeah, I have to constantly remind myself that `end_block` is one past the last too (and it's clumsy to always need to `-1` when printing the `end_block` because users want to know the _last_ block).

But I feel the `begin` and `end` terminology here matches exactly with typical c++ terminology: when calling `begin()` or `end()` for a container you get iterators to the first item and a "one-past-the-end item".

I will add a comment.

Reply:
comment added as part of 4681ad2